File: backend/backend/src/main/java/com/agriconnect/service/ProductService.java
package com.agriconnect.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.agriconnect.entity.Product;
import com.agriconnect.repository.ProductRepository;

@Service
public class ProductService {

    @Autowired
    private ProductRepository productRepository;

    // Add Product
    public Product addProduct(Product product) {

        if (product.getStock() == null || product.getStock() < 0) {
            product.setStock(0);
        }

        product.setInStock(product.getStock() > 0);

        return productRepository.save(product);
    }

    // Get All Products
    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    // Get Products By Retailer
    public List<Product> getProductsByRetailerId(Integer retailerId) {
        return productRepository.findByRetailerId(retailerId);
    }

    // Get Product By ProductId and RetailerId
    public Product getProductByIdAndRetailerId(Integer productId, Integer retailerId) {
        return productRepository
                .findByProductIdAndRetailerId(productId, retailerId)
                .orElse(null);
    }

    // Get Product By ID
    public Product getProductById(Integer id) {
        return productRepository.findById(id).orElse(null);
    }

    // Update Product
    public Product updateProduct(Product product) {

        if (product.getStock() == null || product.getStock() < 0) {
            product.setStock(0);
        }

        product.setInStock(product.getStock() > 0);

        return productRepository.save(product);
    }

    // Delete Product
    public void deleteProduct(Integer id) {
        productRepository.deleteById(id);
    }

    // Search Products
    public List<Product> searchProducts(String keyword) {
        return productRepository.findByNameContainingIgnoreCase(keyword);
    }

    // Filter By Category
    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategoryIgnoreCase(category);
    }

    // Search by Product Name
    public List<Product> searchByProductName(String productName) {
        return productRepository.findByNameContainingIgnoreCase(productName);
    }

    // Filter by Category
    public List<Product> filterByCategory(String category) {
        return productRepository.findByCategoryIgnoreCase(category);
    }

    // Search by Product Name and Category
    public List<Product> searchByProductNameAndCategory(String productName, String category) {
        return productRepository.findByNameContainingIgnoreCaseAndCategoryIgnoreCase(productName, category);
    }
    
    public Product updateStock(Integer productId, Integer retailerId, Integer stock) {

        Product product = productRepository
                .findByProductIdAndRetailerId(productId, retailerId)
                .orElseThrow(() -> new RuntimeException("Product not found"));

        if (stock < 0) {
            throw new RuntimeException("Stock cannot be negative");
        }

        product.setStock(stock);
        product.setInStock(stock > 0);

        return productRepository.save(product);
    }

}

File: frontend/src/services/productService.js
import API from "./api";

// ===============================
// Upload Product Image
// ===============================
export const uploadProductImage = async (file) => {
  const formData = new FormData();

  formData.append("file", file);

  const response = await API.post(
    "/upload/product-image",
    formData
  );

  return response.data.imageUrl || response.data;
};

// ===============================
// Add Product
// ===============================
export const addProduct = async (product) => {
  const response = await API.post(
    "/products",
    product
  );

  return response.data;
};

// ===============================
// Get All Products
// ===============================
export const getAllProducts = async () => {
  const response = await API.get(
    "/products"
  );

  return response;
};

// ===============================
// Get My Products
// ===============================
export const getMyProducts = async () => {
  const response = await API.get(
    "/products/my-products"
  );

  return response.data;
};

// ===============================
// Get Product By ID
// ===============================
export const getProductById = async (id) => {
  const response = await API.get(
    `/products/${id}`
  );

  return response.data;
};

// ===============================
// Update Product
// ===============================
export const updateProduct = async (id, productData) => {
  const response = await API.put(
    `/products/${id}`,
    productData
  );

  return response.data;
};

// ===============================
// Delete Product
// ===============================
export const deleteProduct = async (id) => {
  const response = await API.delete(
    `/products/${id}`
  );

  return response.data;
};

// ===============================
// Update Product Stock
// ===============================
export const updateProductStock = async (
  productId,
  stock
) => {
  const response = await API.patch(
    `/products/${productId}/stock`,
    {
      stock: Number(stock),
    }
  );

  return response.data;
};


